packages/react-reconciler/src/ReactFiberRootScheduler.js

```diff
@@ -0,0 +1,458 @@
+      }
+    } else {
+      // This root still has work. Keep it in the list.
+      prev = root;
+      if (includesSyncLane) {
+        mightHavePendingSyncWork = true;
+      }
+    }
+    root = next;
+  }
+
+  // At the end of the microtask, flush any pending synchronous work. This has
+  // to come at the end, because it does actual rendering work that might throw.
+  flushSyncCallbacks();
+}
+
+function scheduleTaskForRootDuringMicrotask(
+  root: FiberRoot,
+  currentTime: number,
+): Lane {
+  // This function is always called inside a microtask, or at the very end of a
+  // rendering task right before we yield to the main thread. It should never be
+  // called synchronously.
+  //
+  // This function also never performs React work synchronously; it should
+  // only schedule work to be performed later, in a separate task or microtask.
+
+  // Check if any lanes are being starved by other work. If so, mark them as
+  // expired so we know to work on those next.
+  markStarvedLanesAsExpired(root, currentTime);
+
+  // Determine the next lanes to work on, and their priority.
+  const workInProgressRoot = getWorkInProgressRoot();
+  const workInProgressRootRenderLanes = getWorkInProgressRootRenderLanes();
+  const nextLanes = getNextLanes(
+    root,
+    root === workInProgressRoot ? workInProgressRootRenderLanes : NoLanes,
+  );
+
+  const existingCallbackNode = root.callbackNode;
+  if (
+    nextLanes === NoLanes ||
+    // If this root is currently suspended and waiting for data to resolve, don't
+    // schedule a task to render it. We'll either wait for a ping, or wait to
+    // receive an update.
+    (isWorkLoopSuspendedOnData() && root === workInProgressRoot) ||
+    // We should only interrupt a pending commit if the new update
+    // is urgent.
+    (root.cancelPendingCommit !== null && includesOnlyNonUrgentLanes(nextLanes))
+  ) {
+    // Fast path: There's nothing to work on.
+    if (existingCallbackNode !== null) {
+      if (__DEV__ && existingCallbackNode === fakeActCallbackNode) {
+        // Special `act` case: check if this is the fake callback node used by
+        // the `act` implementation.
+      } else {
+        Scheduler_cancelCallback(existingCallbackNode);
+      }
+    }
+    root.callbackNode = null;
+    root.callbackPriority = NoLane;
+    return NoLane;
+  }
+
+  // We use the highest priority lane to represent the priority of the callback.
+  const existingCallbackPriority = root.callbackPriority;
+  const newCallbackPriority = getHighestPriorityLane(nextLanes);
+
+  // Schedule a new callback in the host environment.
+  if (includesSyncLane(newCallbackPriority)) {
+    // Synchronous work will be flushed at the end of the microtask; we don't
+    // need to schedule anything extra.
+    mightHavePendingSyncWork = true;
+    root.callbackPriority = newCallbackPriority;
+    root.callbackNode = null;
+  } else {
+    if (
+      newCallbackPriority === existingCallbackPriority &&
```

Comment:
This moved after the `includesSyncLane` check compared to the old version. Does this condition no longer apply when includesSyncLane?

Reply:
It didn't apply previously either because sync tasks don't get canceled, so I moved it into this more specific branch

Reply:
Doesn't this mean that when we have a task scheduled, and then sync work comes in, we wont cancel the first task anymore?

Reply:
Yeah that's true, though canceling the callback is really just a perf optimization in that case because the extra callback will immediately exit. Though I suppose that's reason enough to move this back to the common path.

Reply:
Ended up adding a separate cancelCallback in that path since it doesn't need the priority check